hazelcast-client/src/main/java/com/hazelcast/client/proxy/ClientPNCounterProxy.java

```diff
@@ -0,0 +1,399 @@
+/*
+ * Copyright (c) 2008-2018, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.client.proxy;
+
+import com.hazelcast.client.impl.protocol.ClientMessage;
+import com.hazelcast.client.impl.protocol.codec.PNCounterAddCodec;
+import com.hazelcast.client.impl.protocol.codec.PNCounterGetCodec;
+import com.hazelcast.client.impl.protocol.codec.PNCounterGetConfiguredReplicaCountCodec;
+import com.hazelcast.client.spi.ClientContext;
+import com.hazelcast.client.spi.ClientProxy;
+import com.hazelcast.cluster.memberselector.MemberSelectors;
+import com.hazelcast.core.HazelcastException;
+import com.hazelcast.core.Member;
+import com.hazelcast.crdt.ConsistencyLostException;
+import com.hazelcast.crdt.TargetNotReplicaException;
+import com.hazelcast.crdt.VectorClock;
+import com.hazelcast.crdt.pncounter.PNCounter;
+import com.hazelcast.internal.util.ThreadLocalRandomProvider;
+import com.hazelcast.nio.Address;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map.Entry;
+import java.util.concurrent.atomic.AtomicReferenceFieldUpdater;
+
+/**
+ * Client proxy implementation for a {@link PNCounter}.
+ */
+public class ClientPNCounterProxy extends ClientProxy implements PNCounter {
+
+    /**
+     * Atomic field updater for the observed clock field
+     */
+    private static final AtomicReferenceFieldUpdater<ClientPNCounterProxy, VectorClock> OBSERVED_TIMESTAMPS_UPDATER =
+            AtomicReferenceFieldUpdater.newUpdater(ClientPNCounterProxy.class, VectorClock.class, "observedClock");
+    private volatile Address currentTargetReplicaAddress;
+    private final Object targetSelectionMutex = new Object();
+    private final Object maxConfiguredReplicaCountMutex = new Object();
```

Comment:
I think we can survive without this mutex. even if max configured replica count is fetched from hz servers multiple times, they will all return the same value.

Reply:
True. I wanted to avoid different client threads retrieving the same config but perhaps it's not such an issue.